akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -139,32 +225,80 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
         case x => receivedUnexpectedFrame(x)
       }
 
+    def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = throw new IllegalStateException(s"pullNextFrame not supported in state $stateName")
+
     /** Called to cleanup any state when the connection is torn down */
-    def shutdown(): IncomingStreamState
+    def shutdown(): Unit = ()
   }
-  case object Idle extends IncomingStreamState {
-    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = SendingData(stream)
-    override def shutdown(): IncomingStreamState = Closed
+
+  case object Idle extends StreamState {
+    def handle(event: StreamFrameEvent): StreamState = expectIncomingStream(event, HalfClosedRemoteWaitingForOutgoingStream(0), OpenReceivingDataFirst(_, 0))
+    override def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = OpenSendingData(outStream, correlationAttributes)
+    override def handleOutgoingCreatedAndFinished(correlationAttributes: Map[AttributeKey[_], _]): StreamState = HalfClosedLocalWaitingForPeerStream(correlationAttributes)
+  }
+  case class OpenReceivingDataFirst(buffer: IncomingStreamBuffer, extraInitialWindow: Int = 0) extends ReceivingData(HalfClosedRemoteWaitingForOutgoingStream(extraInitialWindow)) {
```

Comment:
In some states, we need to track `extraInitialWindow` because window tracking is usually done in OutStream but in some states we don't have an OutStream yet so we need to track it manually until the OutStream is created later.